### PR TITLE
feat: Improve rendering performance on breakpoint changes and re-renders

### PR DESCRIPTION
**Pull request by omulet**
### PR Goal 

First of all, Thanks to all involved in creating this awesome library! The goal of this PR is to improve the overall performance.

### Issues with current implementation

Currently, every single restyled components spawn a new Dimensions Change Listener. This is very expensive especially on react-native-web because window.resize event is fired many times when the browser gets resized. On mobile, the event is basically triggered once for each orientation change, but still there are too many unnecessary listeners.

### Implementation

I have create a `DimensionsProvider` that will pass the dimensions to all components via React Context. This provider is rendered automatically as a child of `<ThemeProvider>`. So the current API is not changed, no breaking changes.

The `useDimensions` hook will read the data from context, instead of spawning its own listener. This way, there is basically 1 listener per each use of ThemeProvider.

In addition, `useRestyle` hook is also better optimised for re-renders. Previously, only the `composedRestyleFunction` was memoized, but the rest of operations for getting the final result were still executed each time the hook was called. Now, the entire result gets memoized and recalculated only if essential deps changes.

### Tests

All existing tests passed without changes. Exception here is `useRestyle.test` which is not complete, maybe it is still in development?

### CODE REVIEW

**Comment by omulet** [comment]
CLA signed, please rerun the check

**Comment by nandorojo** [comment]
Rather than pass the height and width via context, what if you only pass the breakpoint down? That way, components only re-render when the breakpoint changes. This is what I do for [Dripsy](https://dripsy.xyz)

**Comment by omulet** [comment]
@nandorojo this is a good idea, it would require some more changes but it will improve the rendering time even more. But before, I just need to know if the maintainers agrees with this PR idea. Unfortunately I've started to use my fork meanwhile :(
Congrats for your nextjsconf performance btw! Nice packaging of everything you create

**Comment by nandorojo** [comment]
> Congrats for your nextjsconf performance btw! Nice packaging of everything you create

Thanks! Hope it's useful. And good luck here, hope you hear back from the maintainers.

**Comment by drmas** [comment]
@omulet yes of course, all contributions are welcomed.
Could you sign the CLA? as it is required before we can merge your PR

**Comment by omulet** [comment]
> CLA signed, please rerun the check

**Review by nelyousfi — APPROVED** [review]

**Pull request merged**
